synapse/logging/opentracing.py

```diff
@@ -641,19 +641,22 @@ def extract_text_map(carrier):
 # Tracing decorators
 
 
-def trace(func=None, operation_name=None):
+def trace(func=None, opname=None):
     """
     Decorator to trace a function.
-    Sets the operation name to that of the function's.
+    Sets the operation name to that of the function's or that given
+    as operation_name. See the module's doc string for usage
+    examples.
     """
 
-    if func and not operation_name:
-        operation_name = func.__name__
-
     def decorator(func):
         if opentracing is None:
             return func
 
+        # Doing this weird assignment thing to get around local variable
+        # referenced before assignment 'bug' raised by checkstyle
```

Comment:
you mean, why can't you just write:

```
if not opname:
    opname = func.__name__
```

checkstyle is right: it's to do with the python scoping rules. the inner function creates a new scope, and you can't write to variables defined in an outer scope. So by doing the above you are creating a *new* `opname` inside `decorator`, which would only be defined when the *original* opname  was None.

so you're doing the right thing here, though I'd be inclined to rename `operation_name` to `_opname` or `span_name` or something. Either way, it's not a weird assignment thing and it is an actual 'bug' :)

Reply:
Cool, just confused the hell out of me.

Reply:
On the other hand I wasn't looking to write to the outer scope in this case but I can see where this can avoid a wider class of common problems.